### PR TITLE
update readme, add client

### PR DESCRIPTION
**Pull request by deep-diver**
please note `grpc_client.py` is not linkable until this PR is merged.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
...
 *By: [Chansung Park](https://github.com/deep-diver) and [Sayak Paul](https://github.com/sayakpaul)*
 
-This project shows how to serve a TensorFlow image classification model as RESTful and **gRPC** based services with **TFServing**, Docker, and Kubernetes. The idea is to first create a custom TFServing docker image with a TensorFlow model, and then deploy it on a k8s cluster running on [Google Kubernetes Engine (GKE)](https://cloud.google.com/kubernetes-engine). Also we are using [GitHub Actions](https://github.com/features/actions) to automate all the procedures when a new TensorFlow model is released. 
+This project shows how to serve a TensorFlow image classification model as RESTful and **gRPC** based services with **TFServing**, Docker, and Kubernetes. The idea is to first create a custom TFServing docker image with a TensorFlow model, and then deploy it on a k8s cluster running on [Google Kubernetes Engine (GKE)](https://cloud.google.com/kubernetes-engine). **_We are particularly intersted** in deploying **TFServing via gRPC on GKE** by using **[GitHub Actions](https://github.com/features/actions)** to automate all the procedures when a new TensorFlow model is released._**
```
Comment (sayakpaul) [inline README.md]: > We are particularly interested in deploying the model as **_a gRPC client with TF Serving on a k8s cluster using GKE_** and also with GitHub Actions to automate all the procedures when a new TensorFlow model is released.

How about the above?

File: README.md [hunk 2]
```diff
@@ -46,6 +46,10 @@ tfs-server       LoadBalancer   xxxxxxxxxx     xxxxxxxxxx      8500:30869/TCP,85
 kubernetes       ClusterIP      xxxxxxxxxx     <none>          443/TCP                         160m
 ```
 
+## How to perform gRPC inference 
+
+If you wonder how to perform gRPC inference, please check [grpc_client.py](https://github.com/deep-diver/ml-deployment-k8s-tfserving/tree/main/client_grpc_client.py) out. [TFServing API](https://github.com/tensorflow/serving/tree/master/tensorflow_serving/example) provides handy features to construct protobuf request message via `predict_pb2.PredictRequest()`, and `tf.make_tensor_proto(image)` creates protobuf compatible values from `Tensor` data type.
+
```
Comment (sayakpaul) [inline README.md]: [`grpc_client.py`](https://github.com/deep-diver/ml-deployment-k8s-tfserving/tree/main/client_grpc_client.py) provides code to perform inference with the gRPC client.

File: README.md [hunk 3]
```diff
@@ -75,13 +79,14 @@ From the results above,
...
 - [`--enable_batching`](https://github.com/tensorflow/serving/blob/b5a11f1e5388c9985a6fc56a58c3421e5f78149f/tensorflow_serving/model_servers/main.cc#L75) could give you better performance. However, since TFServing doesn't immediately response to each requests, there is a trade-off.
-- By considering cost trade-off, **our recommendation from the experiment is to choose `2n-8c-16r-interop4` configuration** unless you care about dynamic batching capabilities. Or you can write a similar setup by referencing `2n-8c-16r-interop2-batch` but for smaller machines as well. 
+- By considering cost trade-off, **our recommendation from the experiment is to choose `2n-8c-16r-interop4`(2 Nodes of (8vCPU + 16G RAM)) configuration - 2 replicas of TFServing with --tensorflow_inter_op_parallelism=4** unless you care about dynamic batching capabilities. Or you can write a similar setup by referencing `2n-8c-16r-interop2-batch` but for smaller machines as well.
```
Comment (sayakpaul) [inline README.md:83]: **our recommendation from the _experiments_ is to choose...

> unless you care about dynamic batching capabilities

I don't understand how this part is relevant.

File: client/grpc_client.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import numpy as np
+import tensorflow as tf
+from tensorflow_serving.apis import predict_pb2, prediction_service_pb2_grpc
+
+image = tf.image.decode_jpeg(tf.io.read_file("../cat_224x224.jpg"))
+# uncomment below if you want to preprociess on client side
+# image = tf.image.resize(image, (224, 224))[None, ...]
+
+# $ENDPOINT from LoadBalancer or Ingress
+channel = grpc.insecure_channel("$ENDPOINT:8500")
```
Comment (sayakpaul) [inline client/grpc_client.py:11]: Will this work `$ENDPOINT`?

Or is it better to keep a placeholder called `ENDPOINT` after the imports and put a comment on how it should look like?

**Review by sayakpaul — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (deep-diver) [reply README.md]: sounds better @sayakpaul 

just typo I guess. `a gRPC client` ==> `a gRPC server`

File: README.md (hunk 2 above)
Comment (deep-diver) [reply README.md]: thanks!

File: README.md (hunk 1 above)
Comment (sayakpaul) [reply README.md]: Oh, I have seen people referring to APIs as clients too such as REST clients. Hence.

File: client/grpc_client.py (hunk 4 above)
Comment (deep-diver) [reply client/grpc_client.py:11]: any placeholder can be possible, but I guess many ppl prefer to use `envsubst` which automatically replace `$...` with existing environment variables.

something like this

```bash
$ ENDPOINT=127.0.0.1

$ envsubst < grpc_client.py > grpc_client.py
```

File: README.md (hunk 3 above)
Comment (deep-diver) [reply README.md:83]: We have not experimented batching with moderate VM instances, so I was a bit not confident to say batching works great in `2n-8c-16r` spec. WDYT?

File: client/grpc_client.py (hunk 4 above)
Comment (sayakpaul) [reply client/grpc_client.py:11]: Have we reflected this as a comment in the code and made a note in the README? If not, I think it'd make sense to do that.

File: README.md (hunk 3 above)
Comment (sayakpaul) [reply README.md:83]: Oh okay.

File: README.md (hunk 1 above)
Comment (deep-diver) [reply README.md]: That sounds weird to me (https://www.freecodecamp.org/news/rest-api-tutorial-rest-client-rest-service-and-api-calls-explained-with-code-examples/). Maybe better term should be `a gRCP endpoint`. WDYT?

Comment (sayakpaul) [reply README.md]: Yeah perfect.

File: README.md [hunk 5]
```diff
@@ -46,6 +47,10 @@ tfs-server       LoadBalancer   xxxxxxxxxx     xxxxxxxxxx      8500:30869/TCP,85
 kubernetes       ClusterIP      xxxxxxxxxx     <none>          443/TCP                         160m
 ```
 
+## How to perform gRPC inference 
+
+If you wonder how to perform gRPC inference, [grpc_client.py](https://github.com/deep-diver/ml-deployment-k8s-tfserving/tree/main/client_grpc_client.py) provides code to perform inference with the gRPC client ([grpc_client.py](https://github.com/deep-diver/ml-deployment-k8s-tfserving/tree/main/client_grpc_client.py) contains `$ENDPOINT` placeholder. To replace it with your own endpoint, you can `envsubst < grpc_client.py > grpc_client.py` after defining `ENDPOINT` environment variable). [TFServing API](https://github.com/tensorflow/serving/tree/master/tensorflow_serving/example) provides handy features to construct protobuf request message via `predict_pb2.PredictRequest()`, and `tf.make_tensor_proto(image)` creates protobuf compatible values from `Tensor` data type.
```
Comment (deep-diver) [inline README.md:52]: @sayakpaul 

I updated everything we discussed, and I also added the instruction how to replace `$ENDPOINT` in `grpc_client.py`.

**Review by sayakpaul — APPROVED** [review]

**Pull request merged**